Sources/AWSLambdaRuntimeCore/LambdaLifecycle.swift

```diff
@@ -71,22 +71,37 @@ extension Lambda {
         ///
         /// - note: This method must be called  on the `EventLoop` the `Lifecycle` has been initialized with.
         public func start() -> EventLoopFuture<Void> {
-            assert(self.eventLoop.inEventLoop, "Start must be called on the `EventLoop` the `Lifecycle` has been initialized with.")
+            self.eventLoop.assertInEventLoop()
 
             logger.info("lambda lifecycle starting with \(self.configuration)")
             self.state = .initializing
-            // triggered when the Lambda has finished its last run
-            let finishedPromise = self.eventLoop.makePromise(of: Int.self)
-            finishedPromise.futureResult.always { _ in
-                self.markShutdown()
-            }.cascade(to: self.shutdownPromise)
+
             var logger = self.logger
             logger[metadataKey: "lifecycleId"] = .string(self.configuration.lifecycle.id)
             let runner = Runner(eventLoop: self.eventLoop, configuration: self.configuration)
-            return runner.initialize(logger: logger, factory: self.factory).map { handler in
+
+            let startupFuture = runner.initialize(logger: logger, factory: self.factory)
+            startupFuture.flatMap { handler in
+                // after the startup future has succeeded, we have a handler that we can use
+                // to `run` the lambda.
+                let finishedPromise = self.eventLoop.makePromise(of: Int.self)
                 self.state = .active(runner, handler)
                 self.run(promise: finishedPromise)
-            }
+                return finishedPromise.futureResult.always { _ in
+                    // If the lambda is terminated (e.g. LocalServer shutdown), we make sure
+                    // developers have the chance to cleanup their resources.
+                    do {
+                        try handler.syncShutdown()
```

Comment:
@weissi @tomerd We are on the EventLoop at this point. AHCs `httpClient.syncShutdown()` would be a no go at that point. Where do we want to dispatch to?

Reply:
I don't think the `handler` would expect to be called from outside the `EventLoop` to do the shutdown. I think you need to require the caller to provide an asynchronous shutdown method.

Reply:
Well, you could also document that it'll be called on another thread but that's maybe a bit weird. AHC's `syncShutdown` is thread-safe but I don't think we can expect that from everybody.

Also: If shutdown is synchronous, then often the startup is also synchronous, no?

Reply:
> I think you need to require the caller to provide an asynchronous shutdown method.

I tend to agree this fits better with the rest of APIs

Reply:
@weissi @tomerd @ktoso I addressed that, by making `shutdown` async. Further I went a little ahead and wrapped the `eventLoop` as well as the `logger` into a `ShutdownContext`. To get in line with #126 

lmkwyt